ivc/src/ivc/interpreter.rs

```diff
@@ -913,15 +920,101 @@ where
 {
 }
 
+/// Builds selectors for the IVC circuit.
+pub fn build_selectors<F, const N_COL_TOTAL: usize, const N_CHALS: usize>(
+    domain_size: usize,
+) -> Vec<Vec<F>>
+where
+    F: PrimeField,
+{
+    // 3*N + 6*N+2 + N+1 + 35*N + 5 + 1 + N_CHALS =
+    // 45N + 9 + N_CHALS
+    let mut selectors: Vec<Vec<F>> = vec![vec![F::zero(); domain_size]; N_BLOCKS];
+    let mut curr_row = 0;
+    for _i in 0..3 * N_COL_TOTAL {
+        selectors[0][curr_row] = F::one();
+        curr_row += 1;
+    }
+    for _i in 0..6 * N_COL_TOTAL + 2 {
+        selectors[1][curr_row] = F::one();
+        curr_row += 1;
+    }
+    for _i in 0..N_COL_TOTAL + 1 {
+        selectors[2][curr_row] = F::one();
+        curr_row += 1;
+    }
+    for _i in 0..(35 * N_COL_TOTAL + 5) {
+        selectors[3][curr_row] = F::one();
+        curr_row += 1;
+    }
+    for _i in 0..N_CHALS {
+        selectors[4][curr_row] = F::one();
+        curr_row += 1;
+    }
+    for _i in 0..1 {
+        selectors[5][curr_row] = F::one();
+        curr_row += 1;
+    }
+
+    selectors
+}
+
+/// This function generates constraitns for the whole IVC circuit.
+pub fn constrain_selectors<F, Env>(env: &mut Env)
+where
+    F: PrimeField,
+    Env: ColAccessCap<F, IVCColumn>,
+{
+    for i in 0..N_BLOCKS {
+        // Each selector must have value either 0 or 1.
+        let sel = env.read_column(IVCColumn::BlockSel(i));
+        env.assert_zero(sel.clone() * (sel.clone() - Env::constant(F::one())));
+    }
+}
+
+/// This function generates constraitns for the whole IVC circuit.
+pub fn constrain_ivc<F, Ff, Env>(env: &mut Env)
+where
+    F: PrimeField,
+    Ff: PrimeField,
+    Env: ColAccessCap<F, IVCColumn> + LookupCap<F, IVCColumn, IVCLookupTable<Ff>>,
+{
+    constrain_selectors(env);
+
+    let s0 = env.read_column(IVCColumn::BlockSel(0));
+    env.set_assert_mapper(Box::new(move |x| s0.clone() * x));
+    constrain_inputs(env);
+
+    // TODO FIXME add constraints for hashes
+
+    let s2 = env.read_column(IVCColumn::BlockSel(2));
+    env.set_assert_mapper(Box::new(move |x| s2.clone() * x));
```

Comment:
IIUC, it will, for each constraint in constraint_ecadds, add a selector in front of the whole expression, increasing by one the degree?